bd-state/src/lib.rs

```diff
@@ -0,0 +1,287 @@
+    // clearing by prefix in the underlying store, rather than iterating and removing individual
+    // keys.
+    // Ignore errors during clearing - we'll proceed with whatever state we have.
+    let _ = store.clear(Scope::FeatureFlag).await;
+    let _ = store.clear(Scope::GlobalState).await;
+
+    Ok((store, data_loss, previous_snapshot))
+  }
+
+  pub async fn insert(&self, scope: Scope, key: &str, value: String) -> anyhow::Result<()> {
+    let namespaced_key = format!("{}{key}", scope.as_prefix());
+    self
+      .inner
+      .write()
+      .await
+      .insert(
+        namespaced_key,
+        StateValue {
+          value_type: Value_type::StringValue(value).into(),
+          ..Default::default()
+        },
+      )
+      .await
+      .map(|_| ())
+  }
+
+  pub async fn remove(&self, scope: Scope, key: &str) -> anyhow::Result<()> {
+    let namespaced_key = format!("{}{key}", scope.as_prefix());
+    self
+      .inner
+      .write()
+      .await
+      .remove(&namespaced_key)
+      .await
+      .map(|_| ())
+  }
+
+  pub async fn clear(&self, scope: Scope) -> anyhow::Result<()> {
+    let prefix = scope.as_prefix();
+
+    let mut store = self.inner.write().await;
+    let keys_to_remove: Vec<String> = store
+      .as_hashmap()
+      .keys()
+      .filter(|key| key.starts_with(prefix))
+      .cloned()
+      .collect_vec();
+
+    // TODO(snowp): Ideally we should have built in support for batch deletions in the underlying
+    // store. This leaves us open for partial deletions if something fails halfway through.
+    for key in keys_to_remove {
+      store.remove(&key).await?;
+    }
+    Ok(())
+  }
+
+  /// Iterates over all key-value pairs in a given scope.
+  ///
+  /// Returns an iterator that yields (key, value) tuples where the key has been stripped
+  /// of its scope prefix.
+  pub async fn iter_scope(&self, scope: Scope) -> impl IntoIterator<Item = (String, String)> {
+    let prefix = scope.as_prefix();
+    let prefix_len = prefix.len();
+
+    self
+      .inner
+      .read()
+      .await
+      .as_hashmap()
+      .iter()
+      .filter(move |(key, _)| key.starts_with(prefix))
+      .filter_map(move |(key, timestamped_value)| {
+        let stripped_key = &key[prefix_len ..];
+
+        let value = timestamped_value
+          .value
+          .has_string_value()
+          .then(|| timestamped_value.value.string_value())?;
+
+        // TODO(snowp): With the way we are doing the locking etc we cannot return references
+        // whcih seems suboptimal. Consider changing the API to take a closure to avoid allocations.
+        Some((stripped_key.to_string(), value.to_string()))
+      })
+      .collect_vec()
+  }
+
+  pub async fn lock_for_read(&self) -> impl StateReader + '_ {
+    ReadLockedStoreGuard {
+      guard: self.inner.read().await,
+    }
+  }
+
+  pub async fn to_scoped_snapshot(
+    &self,
+    scope: Scope,
+  ) -> AHashMap<String, (String, OffsetDateTime)> {
```

Comment:
What's this for? Per other comment could we just have some kind of filter mechanism to avoid copies?